contractcourt/chain_watcher.go

```diff
@@ -30,20 +30,91 @@ const (
 	maxCommitPointPollTimeout = 10 * time.Minute
 )
 
-// LocalUnilateralCloseInfo encapsulates all the informnation we need to act
-// on a local force close that gets confirmed.
+// LocalUnilateralCloseInfo encapsulates all the informnation we need to act on
+// a local force close that gets confirmed.
 type LocalUnilateralCloseInfo struct {
 	*chainntnfs.SpendDetail
 	*lnwallet.LocalForceCloseSummary
 	*channeldb.ChannelCloseSummary
+
+	// CommitSet is the set of known valid commitments at the time the
+	// remote party's commitemnt hit the chain.
+	CommitSet CommitSet
 }
 
-// CooperativeCloseInfo encapsulates all the informnation we need to act
-// on a cooperative close that gets confirmed.
+// CooperativeCloseInfo encapsulates all the informnation we need to act on a
+// cooperative close that gets confirmed.
 type CooperativeCloseInfo struct {
 	*channeldb.ChannelCloseSummary
 }
 
+// RemoteUnilateralCloseInfo wraps the normal UnilateralCloseSummary to couple
+// the CommitSet at the time of channel closure.
+type RemoteUnilateralCloseInfo struct {
+	*lnwallet.UnilateralCloseSummary
+
+	// CommitSet is the set of known valid commitments at the time the
+	// remote party's commitment hit the chain.
+	CommitSet CommitSet
+}
+
+// CommitSet is a collection of the set of known validate commitments at a
+// given instant. If ConfCommitKey is set, then the commitment identified by
+// the HtlcSetKey has hit the chain. This struct will be used to examine all
+// live HTLCs to determine if any additional actions need to be made based on
+// the remote party's commitments.
+type CommitSet struct {
+	// ConfCommitKey if non nil, then the commitment identified by this
+	// HtlcSetKey has hit the chain.
+	ConfCommitKey *HtlcSetKey
+
+	// LocalCommit is the latest valid commitment for the local party.
+	LocalCommit channeldb.ChannelCommitment
+
+	// RemoteCommit is the latest valid commitment for the remote party.
+	RemoteCommit channeldb.ChannelCommitment
+
+	// RemotePendingCommit if non nil, is the pending commitment for the
+	// remote party. This is a commitment that we've sent to them, but have
+	// not yet received a revocation for.
+	RemotePendingCommit *channeldb.ChannelCommitment
```

Comment:
I think in the case of HTLCSets a map is more apt given the type of manipulation we do to interact with the items. We use pointers here to signal a field that may not always be populated. This struct also has the extra `ConfCommitKey` value as well.

Reply:
In this PR, almost in everywhere a specific set is explictly picked (local, remote, remotepending). It seems that only when something comes in through `htlcUpdates`, the manipulation is easier with a map. If combined with my other comment about just storing the htlcs in `channelArbitrator` rather than the full commit tx, I think the code would be simpler, easier to understand.

Reply:
> In this PR, almost in everywhere a specific set is explictly picked (local, remote, remotepending).

There're also areas where pass all 3 and examine 2 of them. The two representations are isomorphic, as a result it's up to preference which of them are used.